twisted/internet/asyncioreactor.py

```diff
@@ -0,0 +1,261 @@
+# -*- test-case-name: twisted.test.test_internet -*-
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+"""
+asyncio-based reactor implementation.
+"""
+
+from __future__ import absolute_import, division
+
+import errno
+
+from zope.interface import implementer
+
+from twisted.logger import Logger
+from twisted.internet.base import DelayedCall
+from twisted.internet.posixbase import (PosixReactorBase, _NO_FILEDESC,
+                                        _ContinuousPolling)
+from twisted.python.log import callWithLogger
+from twisted.internet.interfaces import IReactorFDSet
+
+try:
+    from asyncio import new_event_loop
+except ImportError:
+    raise ImportError("Requires asyncio.")
+
+
+
+class _DCHandle(object):
+    """
+    Wrapper for L{asyncio.Handle} to be used by
+    L{twisted.internet.base.DelayedCall}.
+    """
+    def __init__(self, handle):
+        self.handle = handle
+
+
+    def cancel(self):
+        """
+        Cancel the inner L{asyncio.Handle}.
+        """
+        self.handle.cancel()
+
+
+
+@implementer(IReactorFDSet)
+class AsyncioSelectorReactor(PosixReactorBase):
+    """
+    Reactor running on top of L{asyncio.SelectorEventLoop}.
+    """
+    _asyncClosed = False
+    _log = Logger()
+
+    def __init__(self, eventloop=None):
+
+        if eventloop is None:
+            eventloop = new_event_loop()
+
+        self._asyncioEventloop = eventloop
+        self._writers = {}
+        self._readers = {}
+        self._delayedCalls = set()
+        self._continuousPolling = _ContinuousPolling(self)
+        super().__init__()
+
+
+    def _readOrWrite(self, selectable, read):
+        method = selectable.doRead if read else selectable.doWrite
+
+        if selectable.fileno() == -1:
+            self._disconnectSelectable(selectable, _NO_FILEDESC, read)
+            return
+
+        try:
+            why = method()
+        except Exception as e:
+            why = e
+            self._log.failure(None)
+        if why:
+            self._disconnectSelectable(selectable, why, read)
+
+
+    def addReader(self, reader):
+        if reader in self._readers.keys() or \
+           reader in self._continuousPolling._readers:
+            return
+
+        fd = reader.fileno()
+        try:
+            self._asyncioEventloop.add_reader(fd, callWithLogger, reader,
+                                              self._readOrWrite, reader,
+                                              True)
+            self._readers[reader] = fd
+        except IOError as e:
```

Comment:
There's a bug in the standard library's `selectors` module that prevents that stdio test from passing.  [I put the gory details into trac](https://twistedmatrix.com/trac/ticket/8367#comment:7), but you can get it to pass by adding the following at the beginning of this `IOError` block and the one in `addWriter`:

```python
self._asyncioEventloop._selector.unregister(fd)
```